src/librustc_mir/interpret/machine.rs

```diff
@@ -219,17 +229,48 @@ pub trait Machine<'mir, 'tcx>: Sized {
         Ok(())
     }
 
-    /// Called for *every* memory access to determine the real ID of the given allocation.
-    /// This provides a way for the machine to "redirect" certain allocations as it sees fit.
+    /// Called for *every* memory access to determine the real ID of the given
+    /// allocation. This provides a way for the machine to "redirect" certain
+    /// allocations as it sees fit.
     ///
-    /// This is used by Miri to redirect extern statics to real allocations.
+    /// This is used by Miri for two purposes:
+    /// 1.  Redirecting extern statics to real allocations.
+    /// 2.  Creating unique allocation ids for thread locals.
+    ///
+    ///     In Rust, one way for creating a thread local is by marking a static
+    ///     with `#[thread_local]`. On supported platforms this gets translated
+    ///     to a LLVM thread local. The problem with supporting these thread
+    ///     locals in Miri is that in the internals of the compiler they look as
+    ///     normal statics, except that they have the `thread_local` attribute.
+    ///     However, in Miri we want to have a property that each allocation has
+    ///     a unique id. Therefore, for these thread locals in
```

Comment:
This sounds like Miri is somehow particular here. It's not, the same happens on actual codegen -- each allocation has a unique address.

Could you reword this more along the lines of, "accessing such a `thread_local` static is actual an effectful computation that depends on the current thread", or so? That gets much more to the point, IMO. That's also why these really should be represented differently, so please also reference https://github.com/rust-lang/rust/issues/70685.